### ISSUE TITLE
Compilation error when using type-generic proc in type definition

### ISSUE THREAD

**Issue by rotu** [nim-lang/Nim#20514]
### What happened?

Using `prod` in a type definition causes the compiler to fail with an uninformative error message.

```nim
import std/math

type Foo[S:static[array[2, int]]] = object
  values: array[prod(S), float]

echo sizeof(Foo[[4,5]])
```

Curiously, replacing `prod` with `sum` causes a different outcome. The below program erroneously prints `1`:

```nim
import std/math

type Foo[S:static[array[2, int]]] = object
  values: array[sum(S), float]

echo sizeof(Foo[[4,5]])
```

### Nim Version

Nim Compiler Version 1.7.3 [Windows: amd64]
Compiled at 2022-10-07
Copyright (c) 2006-2022 by Andreas Rumpf

active boot switches: -d:release

### Current Standard Output Logs

```shell
C:\Users\dan\.choosenim\toolchains\nim-#devel\lib\pure\math.nim(357, 3) Error: cannot generate VM code for result
```

**Comment by rotu** [comment]
possibly addressed by #20423?

**Comment by beef331** [comment]
It's not ideal but a common workaround is to use a template to intercept the typed code so it can actually compile as such:
```nim
import std/math

template arr(a: typed, b: typedesc): auto = array[a, b]

type Foo[S:static[array[2, int]]] = object
  values: arr(prod(S), float)

echo sizeof(Foo[[4,5]])
```

**Comment by rotu** [comment]
@beef331 Thanks! Any insight as to *why* the compiler chokes in the first place?

I wound up wrapping `prod` instead, which seems to also work:
```nim
proc intProd(s:seq[int]):int = prod[int](s)

type Foo[S:static[array[2, int]]] = object
  values: array[intProd(S), float]
```

**Comment by bung87** [comment]
> possibly addressed by #20423?

this one may not related to my PR, that one main problem is inner generic doesn't get `static T` as outter gets, it only get `T`

**Comment by metagn** [comment]
Related #19916 (generic object, static parameter, array index)
